### PR TITLE
Add support for /projects/*/builds/*

### PR DESCRIPTION
**Pull request by ArkBriar**
PR for issue #42

### CODE REVIEW

File: builds.go [hunk 1]
```diff
@@ -0,0 +1,431 @@
...
+	case BuildCanceled:
+		return "canceled"
+	default:
+		return "null"
+	}
+}
+
+// ListBuildsOptions are options for two list apis
+type ListBuildsOptions struct {
+	Scope *[]string `url:"scope, omitempty", json:"scope, omitempty"`
```
Comment (svanharmelen) [inline builds.go]: A slice is already a referenced type (with its zero value being `nil`), so please do not make this a pointer. 

Additionally please leave out the spaces between labels, so `"scope, omitempty"` should be `"scope,omitempty"`. Having the space here is of course not wrong, but I like to keep things consistent throughout the package :wink:

File: builds.go [hunk 2]
```diff
@@ -0,0 +1,431 @@
...
+// GitLab API docs: http://docs.gitlab.com/ce/api/builds.html
+type BuildsService struct {
+	client *Client
+}
+
+// Build represents a ci build.
+//
+// GitLab API docs: http://docs.gitlab.com/ce/api/builds.html
+type Build struct {
+	Commit *struct {
```
Comment (svanharmelen) [inline builds.go]: I think we should use the existing `Commit` struct here: https://github.com/xanzy/go-gitlab/blob/master/commits.go#L36

File: builds.go [hunk 3]
```diff
@@ -0,0 +1,431 @@
...
+		Description string `json:"description"`
+		Active      bool   `json:"active"`
+		IsShared    bool   `json:"is_shared"`
+		Name        string `json:"Name"`
+	} `json:"runner, omitempty"`
+	Stage     string     `json:"stage"`
+	StartedAt *time.Time `json:"started_at"`
+	Status    string     `json:"status"`
+	Tag       bool       `json:"tag"`
+	User      *struct {
```
Comment (svanharmelen) [inline builds.go]: I think we should use the existing `User` struct here: https://github.com/xanzy/go-gitlab/blob/master/users.go#L36

File: builds.go [hunk 4]
```diff
@@ -0,0 +1,431 @@
...
+	Commit *struct {
+		AuthorEmail string     `json:"author_email"`
+		AuthorName  string     `json:"author_name"`
+		CreatedAt   *time.Time `json:"created_at"`
+		ID          string     `json:"id"`
+		Message     string     `json:"message"`
+		ShortID     string     `json:"short_id"`
+		Title       string     `json:"title"`
+	} `json:"commit"`
+	Coverage      *string    `json:"coverage, omitempty"`
```
Comment (svanharmelen) [inline builds.go]: Please do not use a pointer here

File: builds.go [hunk 5]
```diff
@@ -0,0 +1,431 @@
...
+	CreatedAt     *time.Time `json:"created_at"`
+	ArtifactsFile *struct {
```
Comment (svanharmelen) [inline builds.go]: Please do not use a pointer here

File: builds.go [hunk 6]
```diff
@@ -0,0 +1,431 @@
...
+		Filename string `json:"filename"`
+		Size     int    `json:"size"`
+	} `json:"artifacts_file"`
+	FinishedAt *time.Time `json:"finished_at"`
+	ID         int        `json:"id"`
+	Name       string     `json:"name"`
+	Ref        string     `json:"ref"`
+	Runner     *struct {
```
Comment (svanharmelen) [inline builds.go]: Please do not use a pointer here

File: builds.go [hunk 7]
```diff
@@ -0,0 +1,431 @@
...
+		return nil, nil, err
+	}
+	u := fmt.Sprintf("projects/%s/builds", url.QueryEscape(project))
+
+	req, err := s.client.NewRequest("GET", u, opts)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	builds := []Build{}
```
Comment (svanharmelen) [inline builds.go]: `var builds []Build` would be nicer here IMHO (goes for the other locations in this file as well of course)

File: builds.go [hunk 8]
```diff
@@ -0,0 +1,431 @@
...
+//
+// GitLab API docs:
+// http://docs.gitlab.com/ce/api/builds.html#get-a-single-build
+func (s *BuildsService) GetSingleBuild(pid interface{}, buildID int) (*Build, *Response, error) {
+	project, err := parseID(pid)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	u := fmt.Sprintf("projects/%s/builds/%d", project, buildID)
```
Comment (svanharmelen) [inline builds.go]: _nitting_ There is a whiteline above this line, were in all other places the whiteline is below this line...

File: builds.go [hunk 9]
```diff
@@ -0,0 +1,431 @@
...
+	req, err := s.client.NewRequest("GET", u, nil)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	build := &Build{}
```
Comment (svanharmelen) [inline builds.go]: In the rest of this package this is written as `build := new(Build)` (again for consistency)

File: builds.go [hunk 10]
```diff
@@ -0,0 +1,431 @@
...
+//
+// GitLab API docs:
+// http://docs.gitlab.com/ce/api/builds.html#get-build-artifacts
+func (s *BuildsService) GetBuildArtifacts(pid interface{}, buildID int) (io.Reader, *Response, error) {
+	project, err := parseID(pid)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	u := fmt.Sprintf("projects/%s/builds/%d/artifacts", project, buildID)
```
Comment (svanharmelen) [inline builds.go]: Same here about the whiteline...

File: gitlab.go [hunk 11]
```diff
@@ -134,6 +134,7 @@ type Client struct {
 	Settings        *SettingsService
 	SystemHooks     *SystemHooksService
 	Users           *UsersService
+	Builds          *BuildsService
```
Comment (svanharmelen) [inline gitlab.go]: _nitting_ Could you please insert the new `Builds` service following alphabetic ordering?

File: gitlab.go [hunk 12]
```diff
@@ -190,6 +191,7 @@ func newClient(httpClient *http.Client, tokenType tokenType, token string) *Clie
 	c.Settings = &SettingsService{client: c}
 	c.SystemHooks = &SystemHooksService{client: c}
 	c.Users = &UsersService{client: c}
+	c.Builds = &BuildsService{client: c}
```
Comment (svanharmelen) [inline gitlab.go]: _nitting_ Could you please insert the new `c.Builds` service following alphabetic ordering?

File: builds.go [hunk 13]
```diff
@@ -0,0 +1,431 @@
...
+			opts.Scope = &temp
+		}
+	}
+
+	if len(*opts.Scope) == 0 {
+		return nil
+	}
+
+	return opts
+}
```
Comment (svanharmelen) [inline builds.go]: I must admit I don't really like this approach (sorry)... I would leave out the `buildScope ` and `BuildListBuildsOptions ` functions all together and change `ListBuildsOptions` to:

```go
type ListBuildsOptions struct {
  Scope []BuildState `url:"scope,omitempty", json:"scope,omitempty"`
}
```

As `BuildState` is already an existing constant in this package: https://github.com/xanzy/go-gitlab/blob/4f1f8d2b5b095a5e718e3658602598890f4ec746/commits.go#L319

So calling the `ListProjectBuilds` would then become something like:

```go
opts := &ListBuildsOptions{
  Scope: []BuildState{gitlab.Running, gitlab.Pending},
}

builds, resp, err := ListProjectBuilds("svanharmelen/project", opts)
<snip>
```

Which looks pretty user friendly and much cleaner to me...

**Comment by svanharmelen** [comment]
@ArkBriar thanks you so much for this PR!

I left quite some comments, but most of them are very small nits to ensure consistency throughout this package and should just take a minute to update.

Next to that there is one comment (regarding the build scopes) that is about the chosen approach which I would like to change a little to make it more straightforward and (IMO) more logical. But again that should easy enough to update.

But overall it looks great so again thanks for adding this API 😀

**Comment by ArkBriar** [comment]
Thanks for your comments, that makes me more clear about how to write effective go codes.
I have modified builds.go according to your advices and performed an `gofmt` on this package.

The `coverage` field is always null when i test with "gitlab.com" and project "gitlab-org/gitlab-ce".

File: builds.go [hunk 14]
```diff
@@ -0,0 +1,359 @@
...
+	"bytes"
+	"fmt"
+	"io"
+	"net/url"
+	"time"
+)
+
+// ListBuildsOptions are options for two list apis
+type ListBuildsOptions struct {
+	Scope []string `url:"scope,omitempty", json:"scope,omitempty"`
```
Comment (svanharmelen) [inline builds.go]: This should be `Scope []BuildState ...`

File: builds.go [hunk 15]
```diff
@@ -0,0 +1,359 @@
...
+//
+// GitLab API docs: http://docs.gitlab.com/ce/api/builds.html
+type BuildsService struct {
+	client *Client
+}
+
+// Runner represents a ci runner
+//
+// GitLab API docs: http://docs.gitlab.com/ce/api/builds.html
+type Runner struct {
```
Comment (svanharmelen) [inline builds.go]: I don't think we should define `Runner` here, as `Runner` has it's own API: https://docs.gitlab.com/ce/api/runners.html

So until we have implemented that API and it's type, I prefer to use this type as an anonymous embedded struct (like you had before) instead.

File: builds.go [hunk 16]
```diff
@@ -0,0 +1,359 @@
...
+	Description string `json:"description"`
+	Active      bool   `json:"active"`
+	IsShared    bool   `json:"is_shared"`
+	Name        string `json:"name"`
+}
+
+// ArtifactsFile represents an artifacts file (.gitlab.yml)
+//
+// GitLab API docs: http://docs.gitlab.com/ce/api/builds.html
+type ArtifactsFile struct {
```
Comment (svanharmelen) [inline builds.go]: As this is not a known entity I again prefer to also just keep this struct as an embedded struct in the Build struct.

File: builds.go [hunk 17]
```diff
@@ -0,0 +1,359 @@
...
+	Filename string `json:"filename"`
+	Size     int    `json:"size"`
+}
+
+// Build represents a ci build.
+//
+// GitLab API docs: http://docs.gitlab.com/ce/api/builds.html
+type Build struct {
+	Commit        Commit        `json:"commit"`
```
Comment (svanharmelen) [inline builds.go]: Please use a pointer here...

Let me explain a little why this should have a pointer and the previous ones don't. It is not possible to pass around an anonymous struct to another function or variable, so there is no use in it being a pointer. But defined types like `Commit` and `User` can be passed around. So they being a pointer, does make sense :wink:

File: builds.go [hunk 18]
```diff
@@ -0,0 +1,359 @@
...
+	FinishedAt    *time.Time    `json:"finished_at"`
+	ID            int           `json:"id"`
+	Name          string        `json:"name"`
+	Ref           string        `json:"ref"`
+	Runner        Runner        `json:"runner"`
+	Stage         string        `json:"stage"`
+	StartedAt     *time.Time    `json:"started_at"`
+	Status        string        `json:"status"`
+	Tag           bool          `json:"tag"`
+	User          User          `json:"user"`
```
Comment (svanharmelen) [inline builds.go]: Same goes for this one, should be a pointer.

**Comment by svanharmelen** [comment]
@ArkBriar thanks for the updates, just a few additional comments left...

As for the `coverage` , I think it's going to expect a struct here as well. But I also cannot find any detail on that one, so I suggest to just delete the field for now.

Thx!

**Comment by svanharmelen** [comment]
LGTM! Again thanks for the PR @ArkBriar!

**Pull request merged**